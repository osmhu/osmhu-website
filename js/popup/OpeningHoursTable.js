import { addDays, startOfWeek } from 'date-fns';
import OpeningHours from 'opening_hours';

// For opening_hours localization
const nominatimObjectForHungary = {
	class: 'boundary',
	type: 'administrative',
	osm_type: 'relation',
	place_id: '198645167',
	osm_id: '21335',
	address: {
		country: 'Magyarország',
		country_code: 'hu',
	},
};

function nonStopTable() {
	let table = '';
	table += '<table>';
	table += '<tr class="highlighted">';
	table += '<td class="day">Non-stop</td>';
	table += '<td>(Hétfő-vasárnap 0-24)</td>';
	table += '</tr>';
	table += '</table>';
	return table;
}

export default class OpeningHoursTable {
	/*
	 * Highligted ISO day of week: 1 - Monday, 7 - Sunday
	 * Reference: https://date-fns.org/docs/getISODay
	 */
	static generateTable(openingHoursString, highlightedIsoDayOfWeek) {
		let table = '';

		if (openingHoursString === '24/7') {
			return nonStopTable();
		}

		try {
			const days = 'Hétfő Kedd Szerda Csütörtök Péntek Szombat Vasárnap'.split(' ');

			const openingHours = new OpeningHours(openingHoursString, nominatimObjectForHungary);

			table += '<table>';
			let from = startOfWeek(new Date(), { weekStartsOn: 1 });

			for (let day = 0; day < 7; day++) {
				if (day === highlightedIsoDayOfWeek - 1) {
					table += '<tr class="highlighted">';
				} else {
					table += '<tr>';
				}
				table += `<td class="day">${days[day]}</td>`;

				const to = addDays(from, 1);
				const openingHoursOnCurrentDay = [];
				const intervals = openingHours.getOpenIntervals(from, to);
				Object.values(intervals).forEach((interval) => {
					const openFrom = new Date(interval[0]);
					const openTo = new Date(interval[1]);
					const fromHour = openFrom.getHours().toString().padStart(2, '0');
					const fromMinute = openFrom.getMinutes().toString().padStart(2, '0');
					let toHour = openTo.getHours().toString().padStart(2, '0');
					const toMinute = openTo.getMinutes().toString().padStart(2, '0');
					if (toHour === '00' && toMinute === '00') {
						toHour = '24';
					}
					const openInterval = `${fromHour}:${fromMinute} - ${toHour}:${toMinute}`;
					openingHoursOnCurrentDay.push(openInterval);
				});

				if (openingHoursOnCurrentDay.length > 0) {
					table += `<td>${openingHoursOnCurrentDay.join(', ')}</td>`;
				} else {
					table += '<td>zárva</td>';
				}
				from = to;
				table += '</tr>';
			}
			table += '</table>';

			return table;
		} catch (error) {
			throw new Error(`Unable to parse opening hours string '${openingHoursString}': ${error}`);
		}
	}
}
