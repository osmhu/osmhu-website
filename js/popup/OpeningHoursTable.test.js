import OpeningHoursTable from './OpeningHoursTable';

it('throws error for invalid opening hours string', () => {
	expect(() => {
		OpeningHoursTable.generateTable('invalidString', 1);
	}).toThrow();
});

it('renders correctly for non-stop poi', () => {
	const table = OpeningHoursTable.generateTable('24/7', 1);
	expect(table).toMatchSnapshot();
});

it('renders correctly for Mo-Su 10:00-19:00', () => {
	const table = OpeningHoursTable.generateTable('Mo-Su 10:00-19:00', 1);
	expect(table).toMatchSnapshot();
});

it('renders correctly for Mo-Fr 09:00-00:00', () => {
	const table = OpeningHoursTable.generateTable('Mo-Fr 09:00-00:00', 1);
	expect(table).toMatchSnapshot();
});

it('renders correctly for Mo-Sa 07:00-21:00; Su 08:00-18:00', () => {
	const table = OpeningHoursTable.generateTable('Mo-Sa 07:00-21:00; Su 08:00-18:00', 1);
	expect(table).toMatchSnapshot();
});

it('renders correctly for Mo-Fr 09:00-12:00,13:00-17:00; Sa 08:00-12:00', () => {
	const table = OpeningHoursTable.generateTable('Mo-Fr 09:00-12:00,13:00-17:00; Sa 08:00-12:00', 1);
	expect(table).toMatchSnapshot();
});

it('highlights day of the week correctly', () => {
	const table = OpeningHoursTable.generateTable('Mo-Su 10:00-19:00', 3);
	expect(table).toMatchSnapshot();
});
