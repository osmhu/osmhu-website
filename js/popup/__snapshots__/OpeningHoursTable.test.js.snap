// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlights day of the week correctly 1`] = `
<table>
  <tr>
    <td class="day">
      Hétfő
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Kedd
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr class="highlighted">
    <td class="day">
      Szerda
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Csütörtök
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Péntek
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szombat
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Vasárnap
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
</table>
`;

exports[`renders correctly for Mo-Fr 09:00-00:00 1`] = `
<table>
  <tr class="highlighted">
    <td class="day">
      Hétfő
    </td>
    <td>
      09:00 - 24:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Kedd
    </td>
    <td>
      09:00 - 24:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szerda
    </td>
    <td>
      09:00 - 24:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Csütörtök
    </td>
    <td>
      09:00 - 24:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Péntek
    </td>
    <td>
      09:00 - 24:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szombat
    </td>
    <td>
      zárva
    </td>
  </tr>
  <tr>
    <td class="day">
      Vasárnap
    </td>
    <td>
      zárva
    </td>
  </tr>
</table>
`;

exports[`renders correctly for Mo-Fr 09:00-12:00,13:00-17:00; Sa 08:00-12:00 1`] = `
<table>
  <tr class="highlighted">
    <td class="day">
      Hétfő
    </td>
    <td>
      09:00 - 12:00, 13:00 - 17:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Kedd
    </td>
    <td>
      09:00 - 12:00, 13:00 - 17:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szerda
    </td>
    <td>
      09:00 - 12:00, 13:00 - 17:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Csütörtök
    </td>
    <td>
      09:00 - 12:00, 13:00 - 17:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Péntek
    </td>
    <td>
      09:00 - 12:00, 13:00 - 17:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szombat
    </td>
    <td>
      08:00 - 12:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Vasárnap
    </td>
    <td>
      zárva
    </td>
  </tr>
</table>
`;

exports[`renders correctly for Mo-Sa 07:00-21:00; Su 08:00-18:00 1`] = `
<table>
  <tr class="highlighted">
    <td class="day">
      Hétfő
    </td>
    <td>
      07:00 - 21:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Kedd
    </td>
    <td>
      07:00 - 21:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szerda
    </td>
    <td>
      07:00 - 21:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Csütörtök
    </td>
    <td>
      07:00 - 21:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Péntek
    </td>
    <td>
      07:00 - 21:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szombat
    </td>
    <td>
      07:00 - 21:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Vasárnap
    </td>
    <td>
      08:00 - 18:00
    </td>
  </tr>
</table>
`;

exports[`renders correctly for Mo-Su 10:00-19:00 1`] = `
<table>
  <tr class="highlighted">
    <td class="day">
      Hétfő
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Kedd
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szerda
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Csütörtök
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Péntek
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Szombat
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
  <tr>
    <td class="day">
      Vasárnap
    </td>
    <td>
      10:00 - 19:00
    </td>
  </tr>
</table>
`;

exports[`renders correctly for non-stop poi 1`] = `
<table>
  <tr class="highlighted">
    <td class="day">
      Non-stop
    </td>
    <td>
      (Hétfő-vasárnap 0-24)
    </td>
  </tr>
</table>
`;
